src/python/TaskWorker/MasterWorker.py

```diff
@@ -524,16 +560,29 @@ def algorithm(self):
         self.restartQueuedTasks()
         self.logger.debug("Master Worker Starting Main Cycle.")
         while not self.STOP:
-            selection_limit = self.config.TaskWorker.task_scheduling_limit
-            if not self._selectWork(limit=selection_limit):
-                self.logger.warning("No tasks selected.")
-            else:
-                self.logger.info("Work selected successfully.")
+            is_canary = self.config.TaskWorker.is_canary
+            canary_name = self.config.TaskWorker.canary_name
             limit = self.slaves.queueableTasks()
-            if not self._lockWork(limit=limit, getstatus='NEW', setstatus='HOLDING'):
-                time.sleep(self.config.TaskWorker.polling)
-                continue
 
+            # _selectWork and _lockWork are run only if TW is master (not canary)
+            if not is_canary:
+                self.logger.info("This is master TW %s running.", self.config.TaskWorker.name)
+                selection_limit = self.config.TaskWorker.task_scheduling_limit
+                if not self._selectWork(limit=selection_limit):
+                    self.logger.warning("No tasks selected.")
+                else:
+                    self.logger.info("Work selected successfully.")
+                if not self._lockWork(limit=limit, getstatus='NEW', setstatus='HOLDING'):
+                    time.sleep(self.config.TaskWorker.polling)
```

Comment:
hmm.. we need a polling cycle also in the canary TW !!

Reply:
Ok